File: practice/40_Combination_Sum_II.kt
fun main() {
	println(comsum(intArrayOf(2,5,2,1,2), 5))
}

fun comsum(candidates: IntArray, target: Int): List<List<Int>> {
	candidates.sort()
	var result = mutableListOf<List<Int>>()
	helpercom(candidates, target, result, mutableListOf(), 0)
	return result
}

fun helpercom(candidates: IntArray, target: Int, result: MutableList<List<Int>>, temp: MutableList<Int>, start: Int) {
	if(target == 0) {
		result.add(ArrayList(temp))
		return
	}

	if(target < 0) {
		return
	}

	for(i in start until candidates.size) {
		if(i > start && candidates[i] == candidates[i-1]) {
			continue;
		}
		var t = target - candidates[i]
		temp.add(candidates[i])
		helpercom(candidates, t, result, temp, i+1)
		temp.removeAt(temp.size-1)
	}
}

File: practice/40_Combination_Sum_II.js
function sum(candidates, target) {
    candidates.sort((a, b) => a - b);
    let result = [];
    helper(candidates, target, result, [], 0);
    return result;
}

function helper(candidates, target, result, temp, start) {
    if (target === 0) {
        result.push([...temp]);
        return;
    }

    if (target < 0) {
        return;
    }

    for (let index = start; index < candidates.length; index++) {
        if (index > start && candidates[index] === candidates[index - 1]) {
            continue;
        }
        const element = target - candidates[index];
        temp.push(candidates[index]);
        helper(candidates, element, result, temp, index + 1);
        temp.pop();
    }
}

console.log(sum([2, 5, 2, 1, 2], 5));
